src/Workspaces/Core/Portable/Workspace/Solution/LoadableTextAndVersionSource.cs

```diff
@@ -16,12 +17,26 @@ internal sealed class LoadableTextAndVersionSource : ITextAndVersionSource
     private sealed class LazyValueWithOptions
     {
         public readonly LoadableTextAndVersionSource Source;
-        public readonly AsyncLazy<TextAndVersion> LazyValue;
         public readonly LoadTextOptions Options;
 
+        private readonly SemaphoreSlim _gate = new(initialCount: 1);
```

Comment:
❓ Why are we reimplementing another `AsyncLazy<T>` here when we already have two?

Reply:
I'm only aware of one AsyncLazy that we have in roslyn.  We're attempting to simplify that single version we own as it has complexity in it that is only used in this single place, whereas the many dozens of other places do not use that complexity.  We don't think it's worth baking in that single use case into AsyncLazy, especially as it makes it harder to reason about for work we want to do with AsyncLazy in an upcoming PR.

For this special type, we're ok having specialized logic contained within.

Reply:
The savings is on the mental model around AsyncLazy in teh first place, as well as what new ValueSource operations might have to mean if they're on a non-caching vs caching mode.  This effectively just eliminates that entire realm of possibilities from teh data model.

Put another way, this code is the odd code out.  It does something special and means that thinking about async lazy is now always more complex.

Reply:
>  I was OK with "if we're in not caching mode, just don't support the new changes"; 

Sure.  But this also means now extra mental complexity of: do the cases where we ever want to support the new concept ever overlap this code here.